File: payRegister.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.io.IOException;

public class PayRegisterCalculator {

    // Data structures for the three types of records.
    public static class PayrollRecord {
        String employeeID;
        String employeeName;
        String jobTitle;
        String payPeriod;
        double hourlyRate;
        public PayrollRecord(String employeeID, String employeeName, String jobTitle, String payPeriod, double hourlyRate) {
            this.employeeID = employeeID;
            this.employeeName = employeeName;
            this.jobTitle = jobTitle;
            this.payPeriod = payPeriod;
            this.hourlyRate = hourlyRate;
        }
    }

    public static class TimeRecord {
        String employeeID;
        String payPeriod;
        int regularHours;
        int overtimeHours;
        public TimeRecord(String employeeID, String payPeriod, int regularHours, int overtimeHours) {
            this.employeeID = employeeID;
            this.payPeriod = payPeriod;
            this.regularHours = regularHours;
            this.overtimeHours = overtimeHours;
        }
    }

    public static class BenefitsRecord {
        String employeeID;
        String payPeriod;
        double healthInsurance;
        double retirement;
        double otherBenefits;
        public BenefitsRecord(String employeeID, String payPeriod, double healthInsurance, double retirement, double otherBenefits) {
            this.employeeID = employeeID;
            this.payPeriod = payPeriod;
            this.healthInsurance = healthInsurance;
            this.retirement = retirement;
            this.otherBenefits = otherBenefits;
        }
    }

    // Utility: Create a join key from EmployeeID and PayPeriod.
    public static String makeKey(String employeeID, String payPeriod) {
        return employeeID + "|" + payPeriod;
    }

    // Parse one line from payroll_data.csv into a PayrollRecord.
    public static PayrollRecord parsePayrollRecord(String line) {
        if (line == null) return null;
        String[] tokens = line.split(",");
        if (tokens.length < 5) return null;
        String employeeID = tokens[0].trim();
        String employeeName = tokens[1].trim();
        String jobTitle = tokens[2].trim();
        String payPeriod = tokens[3].trim();
        double hourlyRate = Double.parseDouble(tokens[4].trim());
        return new PayrollRecord(employeeID, employeeName, jobTitle, payPeriod, hourlyRate);
    }

    // Parse one line from time_data.csv into a TimeRecord.
    public static TimeRecord parseTimeRecord(String line) {
        if (line == null) return null;
        String[] tokens = line.split(",");
        if (tokens.length < 4) return null;
        String employeeID = tokens[0].trim();
        String payPeriod = tokens[1].trim();
        int regularHours = Integer.parseInt(tokens[2].trim());
        int overtimeHours = Integer.parseInt(tokens[3].trim());
        return new TimeRecord(employeeID, payPeriod, regularHours, overtimeHours);
    }

    // Parse one line from benefits.csv into a BenefitsRecord.
    public static BenefitsRecord parseBenefitsRecord(String line) {
        if (line == null) return null;
        String[] tokens = line.split(",");
        if (tokens.length < 5) return null;
        String employeeID = tokens[0].trim();
        String payPeriod = tokens[1].trim();
        double healthInsurance = Double.parseDouble(tokens[2].trim());
        double retirement = Double.parseDouble(tokens[3].trim());
        double otherBenefits = Double.parseDouble(tokens[4].trim());
        return new BenefitsRecord(employeeID, payPeriod, healthInsurance, retirement, otherBenefits);
    }

    // This method performs a three-way merge join assuming all files are sorted by EmployeeID|PayPeriod.
    public static void mergeJoinAndCompute(String payrollFile, String timeFile, String benefitsFile, String outputFile) {
        long mergeStart = System.currentTimeMillis();
        try (BufferedReader brPayroll = new BufferedReader(new FileReader(payrollFile));
             BufferedReader brTime = new BufferedReader(new FileReader(timeFile));
             BufferedReader brBenefits = new BufferedReader(new FileReader(benefitsFile));
             PrintWriter pw = new PrintWriter(new FileWriter(outputFile))) {

            // Write header for output.
            pw.println("Employee ID,Employee Name,Job Title,Pay Period,Hourly Rate,Regular Hours,Overtime Hours,Gross Wages,Federal Tax,State Tax,Social Security,Medicare,Health Insurance,Retirement,Other Benefits,Total Benefits,Total Deductions,Net Pay");

            // Skip headers in input files.
            String payrollHeader = brPayroll.readLine();
            String timeHeader = brTime.readLine();
            String benefitsHeader = brBenefits.readLine();

            // Read the first record from each file.
            String payrollLine = brPayroll.readLine();
            String timeLine = brTime.readLine();
            String benefitsLine = brBenefits.readLine();

            PayrollRecord currentPayroll = parsePayrollRecord(payrollLine);
            TimeRecord currentTime = parseTimeRecord(timeLine);
            BenefitsRecord currentBenefits = parseBenefitsRecord(benefitsLine);

            while (currentPayroll != null && currentTime != null && currentBenefits != null) {
                String keyPayroll = makeKey(currentPayroll.employeeID, currentPayroll.payPeriod);
                String keyTime = makeKey(currentTime.employeeID, currentTime.payPeriod);
                String keyBenefits = makeKey(currentBenefits.employeeID, currentBenefits.payPeriod);

                // If all keys match, compute and write the output record.
                if (keyPayroll.equals(keyTime) && keyPayroll.equals(keyBenefits)) {
                    double grossWages = currentPayroll.hourlyRate * currentTime.regularHours +
                            1.5 * currentPayroll.hourlyRate * currentTime.overtimeHours;
                    double federalTax = 0.12 * grossWages;
                    double stateTax = 0.05 * grossWages;
                    double socialSecurity = 0.062 * grossWages;
                    double medicare = 0.0145 * grossWages;
                    double totalBenefits = currentBenefits.healthInsurance + currentBenefits.retirement + currentBenefits.otherBenefits;
                    double totalDeductions = federalTax + stateTax + socialSecurity + medicare + totalBenefits;
                    double netPay = grossWages - totalDeductions;

                    String outputLine = String.format("%s,%s,%s,%s,%.2f,%d,%d,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f,%.2f",
                            currentPayroll.employeeID, currentPayroll.employeeName, currentPayroll.jobTitle, currentPayroll.payPeriod,
                            currentPayroll.hourlyRate, currentTime.regularHours, currentTime.overtimeHours, grossWages,
                            federalTax, stateTax, socialSecurity, medicare,
                            currentBenefits.healthInsurance, currentBenefits.retirement, currentBenefits.otherBenefits,
                            totalBenefits, totalDeductions, netPay);
                    pw.println(outputLine);

                    // Advance each file.
                    payrollLine = brPayroll.readLine();
                    timeLine = brTime.readLine();
                    benefitsLine = brBenefits.readLine();
                    currentPayroll = parsePayrollRecord(payrollLine);
                    currentTime = parseTimeRecord(timeLine);
                    currentBenefits = parseBenefitsRecord(benefitsLine);
                } else {
                    // Determine the smallest key lexicographically.
                    String minKey = keyPayroll;
                    if (keyTime.compareTo(minKey) < 0) minKey = keyTime;
                    if (keyBenefits.compareTo(minKey) < 0) minKey = keyBenefits;

                    if (minKey.equals(keyPayroll)) {
                        payrollLine = brPayroll.readLine();
                        currentPayroll = parsePayrollRecord(payrollLine);
                    }
                    if (minKey.equals(keyTime)) {
                        timeLine = brTime.readLine();
                        currentTime = parseTimeRecord(timeLine);
                    }
                    if (minKey.equals(keyBenefits)) {
                        benefitsLine = brBenefits.readLine();
                        currentBenefits = parseBenefitsRecord(benefitsLine);
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        long mergeEnd = System.currentTimeMillis();
        System.out.println("Time for merge join and computation: " + (mergeEnd - mergeStart) + " ms");
    }

    public static void main(String[] args) {
        // File names (adjust paths if necessary).
        String payrollFile = "payroll_data.csv";
        String timeFile = "time_data.csv";
        String benefitsFile = "benefits.csv";
        String outputFile = "payroll_register.csv";

        long totalStart = System.currentTimeMillis();
        mergeJoinAndCompute(payrollFile, timeFile, benefitsFile, outputFile);
        long totalEnd = System.currentTimeMillis();
        System.out.println("Total elapsed time: " + (totalEnd - totalStart) + " ms");
        System.out.println("Pay register computed and saved to " + outputFile);
    }
}

File: payRegister.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"
)

// Data structures for the three input files

type PayrollRecord struct {
	EmployeeID   string
	EmployeeName string
	JobTitle     string
	PayPeriod    string
	HourlyRate   float64
}

type TimeRecord struct {
	EmployeeID    string
	PayPeriod     string
	RegularHours  int
	OvertimeHours int
}

type BenefitsRecord struct {
	EmployeeID      string
	PayPeriod       string
	HealthInsurance float64
	Retirement      float64
	OtherBenefits   float64
}

// Structure for the computed pay register

type PayRegister struct {
	EmployeeID      string
	EmployeeName    string
	JobTitle        string
	PayPeriod       string
	HourlyRate      float64
	RegularHours    int
	OvertimeHours   int
	GrossWages      float64
	FederalTax      float64
	StateTax        float64
	SocialSecurity  float64
	Medicare        float64
	HealthInsurance float64
	Retirement      float64
	OtherBenefits   float64
	TotalBenefits   float64
	TotalDeductions float64
	NetPay          float64
}

// makeKey combines EmployeeID and PayPeriod for map keys.
func makeKey(employeeID, payPeriod string) string {
	return employeeID + "|" + payPeriod
}

// readPayrollRecords reads payroll_data.csv and returns a map keyed by EmployeeID|PayPeriod.
func readPayrollRecords(filename string) (map[string]PayrollRecord, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("cannot open payroll file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, fmt.Errorf("cannot read payroll csv: %v", err)
	}

	payrollMap := make(map[string]PayrollRecord)
	// Skip header
	for i, row := range records {
		if i == 0 {
			continue
		}
		if len(row) < 5 {
			continue
		}
		hourlyRate, err := strconv.ParseFloat(row[4], 64)
		if err != nil {
			return nil, fmt.Errorf("error parsing Hourly Rate in row %d: %v", i+1, err)
		}
		rec := PayrollRecord{
			EmployeeID:   row[0],
			EmployeeName: row[1],
			JobTitle:     row[2],
			PayPeriod:    row[3],
			HourlyRate:   hourlyRate,
		}
		key := makeKey(rec.EmployeeID, rec.PayPeriod)
		payrollMap[key] = rec
	}
	return payrollMap, nil
}

// readTimeRecords reads time_data.csv and returns a map keyed by EmployeeID|PayPeriod.
func readTimeRecords(filename string) (map[string]TimeRecord, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("cannot open time file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, fmt.Errorf("cannot read time csv: %v", err)
	}

	timeMap := make(map[string]TimeRecord)
	for i, row := range records {
		if i == 0 {
			continue // skip header
		}
		if len(row) < 4 {
			continue
		}
		regularHours, err := strconv.Atoi(row[2])
		if err != nil {
			return nil, fmt.Errorf("error parsing Regular Hours in row %d: %v", i+1, err)
		}
		overtimeHours, err := strconv.Atoi(row[3])
		if err != nil {
			return nil, fmt.Errorf("error parsing Overtime Hours in row %d: %v", i+1, err)
		}
		rec := TimeRecord{
			EmployeeID:    row[0],
			PayPeriod:     row[1],
			RegularHours:  regularHours,
			OvertimeHours: overtimeHours,
		}
		key := makeKey(rec.EmployeeID, rec.PayPeriod)
		timeMap[key] = rec
	}
	return timeMap, nil
}

// readBenefitsRecords reads benefits.csv and returns a map keyed by EmployeeID|PayPeriod.
func readBenefitsRecords(filename string) (map[string]BenefitsRecord, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("cannot open benefits file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, fmt.Errorf("cannot read benefits csv: %v", err)
	}

	benefitsMap := make(map[string]BenefitsRecord)
	for i, row := range records {
		if i == 0 {
			continue
		}
		if len(row) < 5 {
			continue
		}
		healthInsurance, err := strconv.ParseFloat(row[2], 64)
		if err != nil {
			return nil, fmt.Errorf("error parsing Health Insurance in row %d: %v", i+1, err)
		}
		retirement, err := strconv.ParseFloat(row[3], 64)
		if err != nil {
			return nil, fmt.Errorf("error parsing Retirement in row %d: %v", i+1, err)
		}
		otherBenefits, err := strconv.ParseFloat(row[4], 64)
		if err != nil {
			return nil, fmt.Errorf("error parsing Other Benefits in row %d: %v", i+1, err)
		}
		rec := BenefitsRecord{
			EmployeeID:      row[0],
			PayPeriod:       row[1],
			HealthInsurance: healthInsurance,
			Retirement:      retirement,
			OtherBenefits:   otherBenefits,
		}
		key := makeKey(rec.EmployeeID, rec.PayPeriod)
		benefitsMap[key] = rec
	}
	return benefitsMap, nil
}

// computeRegister computes the pay register by merging the three datasets.
func computeRegister(payrollMap map[string]PayrollRecord, timeMap map[string]TimeRecord, benefitsMap map[string]BenefitsRecord) []PayRegister {
	var registers []PayRegister

	for key, payroll := range payrollMap {
		timeRec, okTime := timeMap[key]
		benefitsRec, okBenefits := benefitsMap[key]
		if !okTime || !okBenefits {
			// Skip if any record is missing.
			continue
		}

		// Compute Gross Wages:
		// GrossWages = HourlyRate * RegularHours + 1.5 * HourlyRate * OvertimeHours
		grossWages := payroll.HourlyRate*float64(timeRec.RegularHours) +
			1.5*payroll.HourlyRate*float64(timeRec.OvertimeHours)

		// Compute Taxes
		federalTax := 0.12 * grossWages
		stateTax := 0.05 * grossWages
		socialSecurity := 0.062 * grossWages
		medicare := 0.0145 * grossWages

		// Total Benefits
		totalBenefits := benefitsRec.HealthInsurance + benefitsRec.Retirement + benefitsRec.OtherBenefits

		// Total Deductions = Taxes + Total Benefits
		totalDeductions := federalTax + stateTax + socialSecurity + medicare + totalBenefits

		// Net Pay
		netPay := grossWages - totalDeductions

		reg := PayRegister{
			EmployeeID:      payroll.EmployeeID,
			EmployeeName:    payroll.EmployeeName,
			JobTitle:        payroll.JobTitle,
			PayPeriod:       payroll.PayPeriod,
			HourlyRate:      payroll.HourlyRate,
			RegularHours:    timeRec.RegularHours,
			OvertimeHours:   timeRec.OvertimeHours,
			GrossWages:      grossWages,
			FederalTax:      federalTax,
			StateTax:        stateTax,
			SocialSecurity:  socialSecurity,
			Medicare:        medicare,
			HealthInsurance: benefitsRec.HealthInsurance,
			Retirement:      benefitsRec.Retirement,
			OtherBenefits:   benefitsRec.OtherBenefits,
			TotalBenefits:   totalBenefits,
			TotalDeductions: totalDeductions,
			NetPay:          netPay,
		}

		registers = append(registers, reg)
	}

	return registers
}

// writeRegister writes the computed pay register to a CSV file.
func writeRegister(registers []PayRegister, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("cannot create output file: %v", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Write header
	header := []string{
		"Employee ID", "Employee Name", "Job Title", "Pay Period", "Hourly Rate",
		"Regular Hours", "Overtime Hours", "Gross Wages", "Federal Tax", "State Tax",
		"Social Security", "Medicare", "Health Insurance", "Retirement", "Other Benefits",
		"Total Benefits", "Total Deductions", "Net Pay",
	}
	if err := writer.Write(header); err != nil {
		return fmt.Errorf("cannot write header: %v", err)
	}

	// Write each record (format numbers to 2 decimals)
	for _, reg := range registers {
		row := []string{
			reg.EmployeeID,
			reg.EmployeeName,
			reg.JobTitle,
			reg.PayPeriod,
			fmt.Sprintf("%.2f", reg.HourlyRate),
			strconv.Itoa(reg.RegularHours),
			strconv.Itoa(reg.OvertimeHours),
			fmt.Sprintf("%.2f", reg.GrossWages),
			fmt.Sprintf("%.2f", reg.FederalTax),
			fmt.Sprintf("%.2f", reg.StateTax),
			fmt.Sprintf("%.2f", reg.SocialSecurity),
			fmt.Sprintf("%.2f", reg.Medicare),
			fmt.Sprintf("%.2f", reg.HealthInsurance),
			fmt.Sprintf("%.2f", reg.Retirement),
			fmt.Sprintf("%.2f", reg.OtherBenefits),
			fmt.Sprintf("%.2f", reg.TotalBenefits),
			fmt.Sprintf("%.2f", reg.TotalDeductions),
			fmt.Sprintf("%.2f", reg.NetPay),
		}
		if err := writer.Write(row); err != nil {
			return fmt.Errorf("cannot write row: %v", err)
		}
	}

	return nil
}

func main() {
	// File names (adjust as needed)
	payrollFile := "payroll_data.csv"
	timeFile := "time_data.csv"
	benefitsFile := "benefits.csv"
	outputFile := "payroll_register.csv"

	// Start total timer.
	totalStart := time.Now()

	// Step 1: Read Input Files
	readStart := time.Now()
	payrollMap, err := readPayrollRecords(payrollFile)
	if err != nil {
		log.Fatalf("Error reading payroll records: %v", err)
	}

	timeMap, err := readTimeRecords(timeFile)
	if err != nil {
		log.Fatalf("Error reading time records: %v", err)
	}

	benefitsMap, err := readBenefitsRecords(benefitsFile)
	if err != nil {
		log.Fatalf("Error reading benefits records: %v", err)
	}
	readDuration := time.Since(readStart)
	fmt.Printf("Time to read input files: %v\n", readDuration)

	// Step 2: Compute the Pay Register
	computeStart := time.Now()
	registers := computeRegister(payrollMap, timeMap, benefitsMap)
	computeDuration := time.Since(computeStart)
	fmt.Printf("Time to compute pay register: %v\n", computeDuration)
	fmt.Printf("Computed %d register records.\n", len(registers))

	// Step 3: Write the Output CSV
	writeStart := time.Now()
	if err := writeRegister(registers, outputFile); err != nil {
		log.Fatalf("Error writing register file: %v", err)
	}
	writeDuration := time.Since(writeStart)
	fmt.Printf("Time to write output file: %v\n", writeDuration)

	// Total elapsed time
	totalDuration := time.Since(totalStart)
	fmt.Printf("Total elapsed time: %v\n", totalDuration)
	fmt.Printf("Pay register computed and saved to %s\n", outputFile)
}
